state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+        Self {
+            event_key_subscriptions: HashMap::new(),
+            subscription_id_to_event_subscription: HashMap::new(),
+            reconfig_subscriptions: HashMap::new(),
+            storage,
+            next_subscription_id: AtomicU64::new(0),
+        }
+    }
+
+    /// Returns an EventNotificationListener that can be monitored for
+    /// subscribed events. If an event key is subscribed to, it means the
+    /// EventNotificationListener will be sent a notification every time an
+    /// event with the matching key occurs on-chain. Note: if the notification
+    /// buffer fills up too quickly, older notifications will be dropped. As
+    /// such, it is the responsibility of the subscriber to ensure notifications
+    /// are processed in a timely manner.
+    pub fn subscribe_to_events(
+        &mut self,
+        event_keys: Vec<EventKey>,
+    ) -> Result<EventNotificationListener, Error> {
+        if event_keys.is_empty() {
+            return Err(Error::CannotSubscribeToZeroEventKeys);
+        }
+
+        let (notification_sender, notification_receiver) =
+            diem_channel::new(QueueStyle::KLAST, EVENT_NOTIFICATION_CHANNEL_SIZE, None);
+
+        // Create a new event subscription
+        let subscription_id = self.get_new_subscription_id();
+        let event_subscription = EventSubscription {
+            subscription_id,
+            notification_sender,
+            event_buffer: vec![],
+        };
+
+        // Store the new subscription
+        if let Some(old_subscription) = self
+            .subscription_id_to_event_subscription
+            .insert(subscription_id, event_subscription)
+        {
+            return Err(Error::UnknownEventSubscription(
+                subscription_id,
+                format!("{:?}", old_subscription),
+            ));
+        }
+
+        // Update the event key subscriptions to include the new subscription
+        for event_key in event_keys {
+            self.event_key_subscriptions
+                .entry(event_key)
+                .and_modify(|subscriptions| {
+                    subscriptions.insert(subscription_id);
+                })
+                .or_insert_with(|| HashSet::from_iter(vec![subscription_id].iter().cloned()));
+        }
+
+        Ok(EventNotificationListener {
+            notification_receiver,
+        })
+    }
+
+    /// Returns a ReconfigNotificationListener that can be monitored for
+    /// reconfiguration events. Subscribers will be sent a notification
+    /// containing all new on-chain configuration values whenever a new epoch
+    /// begins. Note: if the notification buffer fills up too quickly, older
+    /// notifications will be dropped. As such, it is the responsibility of the
+    /// subscriber to ensure notifications are processed in a timely manner.
+    pub fn subscribe_to_reconfigurations(&mut self) -> Result<ReconfigNotificationListener, Error> {
```

Comment:
do we need this special subscribe? reconfiguration has its own event `NewEpochEvent` that can be subscribed. although it doesn't have all the payloads in the event

Reply:
Yeah, I was debating this in my mind for a while, and I ultimately decided to include it because: (i) it provides a cleaner interface (e.g., if a client only cares about reconfigurations and not about events, they can just get what they need and not deal with the messiness of event keys); (ii) the types are clearer, e.g., if we wanted to include all the payloads, we'd need to extend an `EventNotification` to include an optional OnChainConfigPayload.. which isn't great; and (iii) it makes reasoning about the notification logic a little clearer, in my opinion 😄